lib/services/notifications_cache.go

```diff
@@ -0,0 +1,521 @@
+func NewGlobalNotificationCache(cfg NotificationCacheConfig) (*GlobalNotificationCache, error) {
+	if err := cfg.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+
+	ttlCache, err := utils.NewFnCache(utils.FnCacheConfig{
+		Context: ctx,
+		TTL:     15 * time.Second,
+		Clock:   cfg.Clock,
+	})
+	if err != nil {
+		cancel()
+		return nil, trace.Wrap(err)
+	}
+
+	c := &GlobalNotificationCache{
+		cfg:          cfg,
+		ttlCache:     ttlCache,
+		initC:        make(chan struct{}),
+		closeContext: ctx,
+		cancel:       cancel,
+	}
+
+	if _, err := newResourceWatcher(ctx, c, ResourceWatcherConfig{
+		Component: "global-notification-cache",
+		Client:    cfg.Events,
+	}); err != nil {
+		cancel()
+		return nil, trace.Wrap(err)
+	}
+
+	return c, nil
+}
+
+// streamUserNotifications returns a stream with the user-specific notifications in the cache for a specified user sorted from newest to oldest.
+// We use streams here as it's a convenient way for us to construct pages to be returned to the UI one item at a time in combination with global notifications.
+func (c *UserNotificationCache) StreamUserNotifications(ctx context.Context, username string, startKey string) stream.Stream[*notificationsv1.Notification] {
+	if username == "" {
+		return stream.Fail[*notificationsv1.Notification](trace.BadParameter("username is required for fetching user notifications"))
+	}
+
+	cache, err := c.read(ctx)
+	if err != nil {
+		return stream.Fail[*notificationsv1.Notification](trace.Wrap(err))
+	}
+
+	if !cache.HasIndex(notificationKey) {
+		return stream.Fail[*notificationsv1.Notification](trace.Errorf("user notifications cache was not configured with index %q (this is a bug)", notificationKey))
+	}
+
+	if startKey == "" {
+		// If the startKey isn't specified, we get the initial startKey by descending through the user notifications cache until we get the first notification
+		// for this user's username.
+		cache.Descend(notificationKey, "", "", func(notification *notificationsv1.Notification) bool {
+			// If this notification is for the user, we set the startKey to it and exit the traversal.
+			if notification.GetSpec().GetUsername() == username {
+				startKey = GetUserSpecificKey(notification)
+				// Exit traversal.
+				return false
+			}
+
+			return true
+		})
+	}
+
+	notifications := []*notificationsv1.Notification{}
+
+	// We get all the user-specific notifications for this user from the cache up-front, starting with the startKey.
+	cache.Descend(notificationKey, startKey, "", func(n *notificationsv1.Notification) bool {
+		// Once we reach notifications that belong to a different user, it means that this user has no more user-specific notifications and we stop traversing.
+		if n.GetSpec().GetUsername() != username {
+			return false
+		}
+
+		notifications = append(notifications, n)
+		return true
+	})
+
+	return stream.Slice(notifications)
```

Comment:
We generally always clone values that get returned from this kind of cache.  It's too easy to accidentally introduce race conditions otherwise.  The common pattern is to allow the caller to pass in a matcher closure that can tell us in advance which values the caller is actually going to care about, and then clone only those to bubble up to them.  That way be can get the safety of defensive copies without going overboard if the caller doesn't care about all of them.

It's also preferable to have pagination/streaming logic be paginated "all the way down", just so that per-request memory usage doesn't grow unchecked.  Rather than allocating a slice of all notifications in a stream, I'd suggest only gathering up to some fixed size limit per read.  One way to do this would be to have this API be paginated.  Another would be to set up a helper function that gets one page at a time and then passing that into a `stream.PageFunc`, which will convert it into a stream.